### PR TITLE
Add memcached into benchmark and CI

### PR DESCRIPTION
**Pull request by Cautreoxit**
This PR is to add memcached into benchmark and CI. In the next step, Pytorch will be added, too.

This PR lets guest to run a Memcached server, while host running **memaslap**, a benchmarking tool designed to test the performance of Memcached. It currently implements three tests for memcached: 
`"benchmarks": [
        "8_t_128_conc_10k_window",
        "8_t_128_conc_20k_window",
        "16_t_512_conc_10k_window"
    ]`. The meaning of these parameters in the manual are as followed:
- -T: Number of threads to startup, better equal to CPU numbers. Default 8.
- -c: Number of concurrency to simulate with load. Default 128.
- -w: Task window size of each concurrency, suffix: K, M e.g.: --win_size=10k. Default 10k.

There are also other parameters for **memaslap**, so we could easily add other tests in need.

This PR has several places to be reviewed:
1. the structure of files and lines
2. other tests of memcached to add

Now I could successfully run these three tests locally. After everything is ok, I would update the Dockerfile image and bump the version file.

### CODE REVIEW

**Comment by Cautreoxit** [comment]
By the way, I could see these three benchmarks' results locally. They are not very good:

8_t_128_conc_10k_window:
`[
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Linux",
    "unit": "operations per second",
    "value": "3627147",
    "extra": "linux_result"
  },
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Asterinas",
    "unit": "operations per second",
    "value": "1963823",
    "extra": "aster_result"
  }
]
`

8_t_128_conc_20k_window:
`[
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Linux",
    "unit": "operations per second",
    "value": "4202602",
    "extra": "linux_result"
  },
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Asterinas",
    "unit": "operations per second",
    "value": "2143648",
    "extra": "aster_result"
  }
]
`

16_t_512_conc_10k_window:
`[
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Linux",
    "unit": "operations per second",
    "value": "4888952",
    "extra": "linux_result"
  },
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Asterinas",
    "unit": "operations per second",
    "value": "1133348",
    "extra": "aster_result"
  }
]
`

**Comment by tatetian** [comment]
@Cautreoxit Could you add PyTorch and Memcached to the Dockerfile in the same PR (but in two separate commits)? And in that PR, bump the version to trigger Docker image building. After the two applications and their benchmarks are contained in the latest Docker image, you can submit another PR to enable their benchmarks.

**Comment by Cautreoxit** [comment]
> > * -c: Number of concurrency to simulate with load. Default 128.
> 
> @Cautreoxit Did you or could you try to use a smaller number for concurrency? Our current network stack performs well when the number of concurrent connections is low and does poorly when the number is high. This is a known limitation and @StevenJiang1110 is working on it.

Sure. I changed the number for concurrency in two tests:

8_t_64_conc_10k_window:
`[
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Linux",
    "unit": "operations per second",
    "value": "3679629",
    "extra": "linux_result"
  },
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Asterinas",
    "unit": "operations per second",
    "value": "2412752",
    "extra": "aster_result"
  }
]
`

8_t_32_conc_10k_window:
`[
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Linux",
    "unit": "operations per second",
    "value": "1942432",
    "extra": "linux_result"
  },
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Asterinas",
    "unit": "operations per second",
    "value": "2646804",
    "extra": "aster_result"
  }
]`

As we can see, as the number for concurrency goes down, Asterinas's performance versus Linux goes up. We could also reduce the number further. I think we could select the most suitable number of concurrency for our evaluation.

**Comment by Cautreoxit** [comment]
Another information: I run memcached with smp=16 on Asterinas, and it finishes successfully. However, its performance versus Linux goes down:
`[
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Linux",
    "unit": "operations per second",
    "value": "4734185",
    "extra": "linux_result"
  },
  {
    "name": "Average OPS of SET and GET over virtio-net between Host Linux and Guest Asterinas",
    "unit": "operations per second",
    "value": "1013028",
    "extra": "aster_result"
  }
]
`

**Comment by Cautreoxit** [comment]
I add memcached into Dockerfile and bump the version now. If everything is good, after it is merged, I will add memcached into benchmark in another PR.

File: tools/docker/Dockerfile.jinja [hunk 1]
```diff
@@ -111,6 +114,14 @@ WORKDIR /root/schbench
 RUN make \
     && cp schbench /usr/local/benchmark/
 
+# Build memaslap for memcached
+WORKDIR /root/libmemcached-1.0.18
+RUN LDFLAGS='-lpthread' CPPFLAGS='-fcommon -fpermissive' CFLAGS='-fpermissive -fcommon' \
+    ./configure --enable-memaslap --prefix=/usr/local/benchmark/libmemcached \ 
+    && CPPFLAGS='-fcommon' make -j \ 
+    && make install \
+    && ldconfig
```
Comment (tatetian) [inline tools/docker/Dockerfile.jinja:123]: Is the ending `ldconfig` really necessary?

**Review by tatetian — COMMENTED** [review]
Looks most good. Just one minor issue. Thanks!

File: tools/docker/Dockerfile.jinja (hunk 1 above)
Comment (Cautreoxit) [reply tools/docker/Dockerfile.jinja:123]: Oh I think it's not really necessary. Thanks! I will delete it.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**